File: Java/0030_substring_with_concatenation_of_all_words/0030_substring_with_concatenation_of_all_words.java
class Solution {
    public List<Integer> findSubstring(String s, String[] words) {
        if (words.length == 0) {
            return new ArrayList<Integer>();
        }   
        int wordsSize = words[0].length();
        int wordsNum = words.length;
        Map<String, Integer> wordCnt = new HashMap<String, Integer>();
        Map<String, Integer> wordCntTmp = new HashMap<String, Integer>();
        List<Integer> res = new ArrayList<Integer>();
        for (int i = 0; i < wordsNum; i++) {
            if (wordCnt.containsKey(words[i])) {
                int count = wordCnt.get(words[i]) + 1;
                wordCnt.put(words[i], count);
            }
            else {
                wordCnt.put(words[i], 1);
            }
        }
        for (int i = 0; i < s.length() - wordsNum * wordsSize + 1; i++) {
            int j = 0;
            wordCntTmp.clear();
            for (j = 0; j < wordsNum; j++) {
                String word = s.substring(i + j * wordsSize, i + j * wordsSize + wordsSize);
                if (wordCnt.containsKey(word)) {
                    if (wordCntTmp.containsKey(word)) {
                        int count = wordCntTmp.get(word) + 1;
                        wordCntTmp.put(word, count);
                    }
                    else {
                        wordCntTmp.put(word, 1);
                    }
                    if (wordCntTmp.get(word) > wordCnt.get(word)) {
                        break;
                    }
                }
                else {
                    break;
                }
            }
            if (j == wordsNum) {
                res.add(i);
            }
        }
        return res;
    }
}

File: Python/0030_substring_with_concatenation_of_all_words/0030_substring_with_concatenation_of_all_words.py
class Solution:
    def findSubstring(self, s, words):
        """
        :type s: str
        :type words: List[str]
        :rtype: List[int]
        """
        if len(words) == 0:
            return []
        wordCnt = {}
        wordCntTmp = {}
        res = []
        for i in range(len(words)):
            if words[i] in wordCnt:
                wordCnt[words[i]] += 1
            else:
                wordCnt[words[i]] = 1
        for i in range(0, len(s) - len(words[0]) * len(words) + 1):
            j = 0
            wordCntTmp.clear()
            while j < len(words):
                word = s[i + j * len(words[0]) : i + j * len(words[0]) + len(words[0])]
                if word in wordCnt:
                    if word in wordCntTmp:
                        wordCntTmp[word] += 1
                    else:
                        wordCntTmp[word] = 1
                    if wordCntTmp[word] > wordCnt[word]:
                        break
                else:
                    break
                j += 1
            if j == len(words):
                res.append(i)
        return res